src/java.base/share/classes/jdk/internal/classfile/impl/Util.java

```diff
@@ -0,0 +1,234 @@
+                    break;
+                default:
+                    bs.set(i);
+                    if (type.charAt(i) == 'L') {
+                        while (type.charAt(++i) != ';')
+                            ;
+                    }
+            }
+        }
+        return bs;
+    }
+
+    @SuppressWarnings("fallthrough")
+    public static int parameterSlots(String type) {
+        BitSet bs = findParams(type);
+        int count = 0;
+        for (int i = bs.nextSetBit(0); i >= 0; i = bs.nextSetBit(i+1)) {
+            count += (type.charAt(i) == 'J' || type.charAt(i) == 'D') ? 2 : 1;
+        }
+        return count;
+    }
+
+    public static int[] parseParameterSlots(int flags, String type) {
+        BitSet bs = findParams(type);
+        int[] result = new int[bs.cardinality()];
+        int index = 0;
+        int count = ((flags & ACC_STATIC) != 0) ? 0 : 1;
+        for (int i = bs.nextSetBit(0); i >= 0; i = bs.nextSetBit(i+1)) {
+            result[index++] = count;
+            count += (type.charAt(i) == 'J' || type.charAt(i) == 'D') ? 2 : 1;
+        }
+        return result;
+    }
+
+    public static int maxLocals(int flags, String type) {
+        BitSet bs = findParams(type);
+        int count = ((flags & ACC_STATIC) != 0) ? 0 : 1;
+        for (int i = bs.nextSetBit(0); i >= 0; i = bs.nextSetBit(i+1))
+            count += (type.charAt(i) == 'J' || type.charAt(i) == 'D') ? 2 : 1;
+        return count;
+    }
+
+    public static String toClassString(String desc) {
+        //TODO: this doesn't look right L ... ;
+        return desc.replaceAll("/", ".");
+    }
+
+    public static Iterator<String> parameterTypes(String s) {
+        //TODO: gracefully non-method types
+        return new Iterator<>() {
+            int ch = 1;
+
+            public boolean hasNext() {
+                return s.charAt(ch) != ')';
+            }
+
+            public String next() {
+                char curr = s.charAt(ch);
+                switch (curr) {
+                    case 'C', 'B', 'S', 'I', 'J', 'F', 'D', 'Z':
+                        ch++;
+                        return String.valueOf(curr);
+                    case '[':
+                        ch++;
+                        return "[" + next();
+                    case 'L': {
+                        int start = ch;
+                        while (s.charAt(++ch) != ';') { }
+                        ++ch;
+                        return s.substring(start, ch);
+                    }
+                    default:
+                        throw new AssertionError("cannot parse string: " + s);
+                }
+            }
+        };
+    }
+
+    public static String returnDescriptor(String s) {
+        return s.substring(s.indexOf(')') + 1);
+    }
+
+    public static String toInternalName(ClassDesc cd) {
+        var desc = cd.descriptorString();
+        return switch (desc.charAt(0)) {
+            case '[' -> desc;
```

Comment:
Is this correct? Arrays don't have an internal name.

Reply:
It is a workaround to get CP class entry name string from descriptor, so the  `toInternalName` is not exact.
I'll handle the array descriptors separately and let this method to handle real class internal names only.